Add LLVM patch D14260 to improve SIMD code

eschnett opened the pull request:
Arch Robison proposed the patch http://reviews.llvm.org/D14260 "Optimize store of "bitcast" from vector to aggregate" for LLVM. This patch applies cleanly to LLVM 3.7.1. It seems to be the last missing puzzle piece on the LLVM side to allow generating efficient SIMD instructions via `llvm_call` in Julia. For an example package, see e.g. https://github.com/eschnett/SIMD.jl.

Some discussion relevant to this PR are in #5355. @ArchRobison, please comment.

Julia stores tuples as LLVM arrays, whereas LLVM SIMD instructions require LLVM vectors. The respective conversions are unfortunately not always optimized out unless the patch above is applied, leading to a cumbersome sequence of instructions to disassemble and reassemble a SIMD vector. An example is given here https://github.com/eschnett/SIMD.jl/issues/1#issuecomment-180997930.

Without this patch, the loop kernel looks like (x86-64, AVX2 instructions):

```
    vunpcklpd   %xmm4, %xmm3, %xmm3 # xmm3 = xmm3[0],xmm4[0]
    vunpcklpd   %xmm2, %xmm1, %xmm1 # xmm1 = xmm1[0],xmm2[0]
    vinsertf128 $1, %xmm3, %ymm1, %ymm1
    vmovupd 8(%rcx), %xmm2
    vinsertf128 $1, 24(%rcx), %ymm2, %ymm2
    vaddpd  %ymm2, %ymm1, %ymm1
    vpermilpd   $1, %xmm1, %xmm2 # xmm2 = xmm1[1,0]
    vextractf128    $1, %ymm1, %xmm3
    vpermilpd   $1, %xmm3, %xmm4 # xmm4 = xmm3[1,0]
Source line: 62
    vaddsd  (%rcx), %xmm0, %xmm0
```

Note that the SIMD vector is kept in register `%ymm1`, but is unnecessarily scalarized into registers `%xmm{0,1,2,3}` at the end of the kernel, and re-assembled in the beginning.

With this patch, the loop kernel looks like:

```
L192:
    vaddpd  (%rdx), %ymm1, %ymm1
Source line: 62
    addq    %rsi, %rdx
    addq    %rcx, %rdi
    jne L192
```

which is perfect.

tkelman commented on deps/Makefile:
```diff
@@ -658,6 +658,12 @@ $$(LLVM_SRC_DIR)/$1.patch-applied: $(LLVM_SRC_DIR)/configure | $$(SRCDIR)/$1.pat
 endef
+define LLVM_PATCH0
+$$(LLVM_SRC_DIR)/$1.patch-applied: $(LLVM_SRC_DIR)/configure | $$(SRCDIR)/$1.patch
+	cd $$(LLVM_SRC_DIR) && patch -p0 < $$(SRCDIR)/$1.patch
```
why not just adjust the patch content so it can be applied consistently with all the others instead of duplicating application code

eschnett replied:
I wanted to keep the original LLVM patch without modification.

I can, of course, modify it -- but then it might be easier to just append it to one of the existing patches instead of introducing a new one.

tkelman replied:
Modifying the existing patches would lead to the `.patch-applied` files indicating an incorrect state in everyone's existing build, so the new modifications would not get applied except in from-scratch builds of llvm.

tkelman commented on deps/Makefile:
```diff
@@ -671,7 +677,10 @@ else ifeq ($(LLVM_VER),3.7.1)
+$(eval $(call LLVM_PATCH0,llvm-D14260))
 $(LLVM_SRC_DIR)/llvm-3.7.1_2.patch-applied: $(LLVM_SRC_DIR)/llvm-3.7.1.patch-applied
+$(LLVM_SRC_DIR)/llvm-3.7.1_3.patch-applied: $(LLVM_SRC_DIR)/llvm-3.7.1_2.patch-applied
+$(LLVM_SRC_DIR)/llvm-D14260.patch-applied: $(LLVM_SRC_DIR)/llvm-3.7.1_3.patch-applied
```
these are not required unless multiple patches modify the same source files

eschnett replied:
I find it tedious to look through all existing patches, extracting all file names, and comparing against all file names in the new patch, manually determining which patches conflict. It's easy to make a mistake. Applying patches is surely so fast that the missed Makefile parallelization doesn't matter. I think the only reason the patches aren't applied serially is that this is difficult to do in Makefiles, hence the dependencies.

Please advise -- I'm not objecting, I'm just giving the reasons for why I did things the way I did.

tkelman replied:
`cat deps/llvm-3.7.*.patch | grep +++` ? If there's no danger of a race condition, there's no dependency.

eschnett replied:
Very good, I'll remove the dependency. I guess it's the task of the code reviewer to confirm this is correct...

eschnett committed:
Add LLVM patch D14260 to improve SIMD code

Arch Robison proposed the patch <http://reviews.llvm.org/D14260> "Optimize store of "bitcast" from vector to aggregate" for LLVM. This patch applies cleanly to LLVM 3.7.1. It seems to be the last missing puzzle piece on the LLVM side to allow generating efficient SIMD instructions via `llvm_call` in Julia. For an example package, see e.g. <https://github.com/eschnett/SIMD.jl>.

Some discussion relevant to this PR are in #5355. @ArchRobison, please comment.

Julia stores tuples as LLVM arrays, whereas LLVM SIMD instructions require LLVM vectors. The respective conversions are unfortunately not always optimized out unless the patch above is applied, leading to a cumbersome sequence of instructions to disassemble and reassemble a SIMD vector. An example is given here <https://github.com/eschnett/SIMD.jl/issues/1#issuecomment-180997930>.

Without this patch, the loop kernel looks like (x86-64, AVX2 instructions):

```
vunpcklpd %xmm4, %xmm3, %xmm3 # xmm3 = xmm3[0],xmm4[0]
vunpcklpd %xmm2, %xmm1, %xmm1 # xmm1 = xmm1[0],xmm2[0]
vinsertf128 $1, %xmm3, %ymm1, %ymm1
vmovupd 8(%rcx), %xmm2
vinsertf128 $1, 24(%rcx), %ymm2, %ymm2
vaddpd %ymm2, %ymm1, %ymm1
vpermilpd $1, %xmm1, %xmm2 # xmm2 = xmm1[1,0]
vextractf128 $1, %ymm1, %xmm3
vpermilpd $1, %xmm3, %xmm4 # xmm4 = xmm3[1,0]
Source line: 62
vaddsd (%rcx), %xmm0, %xmm0
```

Note that the SIMD vector is kept in register `%ymm1`, but is unnecessarily scalarized into registers `%xmm{0,1,2,3}` at the end of the kernel, and re-assembled in the beginning.

With this patch, the loop kernel looks like:

```
L192:
vaddpd	(%rdx), %ymm1, %ymm1
Source line: 62
addq	%rsi, %rdx
addq	%rcx, %rdi
jne	L192
```

which is perfect.

eschnett commented:
Addressed @tkelman's comments, rebased.
